### ISSUE TITLE
Flip order of s[a..b].as_bytes()

### ISSUE THREAD

**Issue by A-Walrus** [rust-lang/rust-clippy#10981]
### What it does

Suggests replacing slicing a `str` followed by `as_bytes` with the opposite order
```rust
s[a..b].as_bytes()
```
with
```rust
&s.as_bytes()[a..b]
```

### Advantage

- Removes unnecessary utf8 indexing check 
- won't panic if a and b are not on utf8 boundaries

### Drawbacks

Technically possible someone relied on it panicking, but find that unlikely.

### Example

```rust
let s = "Hello World";
let bytes = s[1..3].as_bytes();
```

Could be written as:

```rust
let s = "Hello World";
let bytes = &s.as_bytes()[1..3];
```

**Comment by A-Walrus** [comment]
I'd like to work on this but I'm new to the codebase and might need some guidance.
Want to make sure I'm on the right track... this lint should implement the `LateLintPass` `check_expr` method right?

**Comment by y21** [comment]
This seems like it would need a `LateLintPass`, since you need to know that the type of the expression being indexed is `str` (type information are only available in late lint passes), and `check_expr` sounds right to me. I would implement the `check_expr` method, check if it's a method call to `str::as_bytes` and the receiver of the call is an index expression, with the indexed type being `str`. The [clippy documentation (specifically "common tools")](https://doc.rust-lang.org/clippy/development/common_tools_writing_lints.html) is probably helpful here. Good luck!
